src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Documents/WinRTSpellerInteropExtensions.cs

```diff
@@ -0,0 +1,184 @@
+        ///       problems this heuristic is intended to help with. 
+        ///       
+        ///     - Typical word-breaking problems that have been observed empirically involve only one missed character,
+        ///       for which MAXLEN=1 would be sufficient. MAXLEN=4 is chosen as a sufficiently-large tradeoff between
+        ///       correctness and performance. 
+        ///       
+        ///     - Also see https://github.com/dotnet/wpf/pull/2753#issuecomment-602120768 for a discussion related to this. 
+        /// </remarks>
+        public static IReadOnlyList<SpellerSegment> ComprehensiveGetTokens(
+            this WordsSegmenter segmenter,
+            string text,
+            SpellChecker spellChecker,
+            WinRTSpellerInterop owner)
+        {
+            IReadOnlyList<WordSegment> tokens = segmenter?.GetTokens(text) ?? new List<WordSegment>().AsReadOnly();
+            var allTokens = new List<SpellerSegment>();
+
+            int predictedNextTokenStartPosition = 0;
+            for (int i = 0; i < tokens.Count; i++)
+            {
+                int nextTokenStartPosition = (int)tokens[i].SourceTextSegment.StartPosition;
+
+                if (spellChecker != null)
+                {
+                    if (nextTokenStartPosition > predictedNextTokenStartPosition)
+                    {
+                        // There is a "gap" between the last recorded token and the current token.
+                        // Identify the missing token and add it as a "supplementary word segment" - but only if the token
+                        // turns out to be a substantial one (i.e., if the string is non-blank/non-empty). 
+                        var missingFragment =
+                            new SpellerSegment(
+                                text,
+                                new WinRTSpellerInterop.TextRange(
+                                    predictedNextTokenStartPosition,
+                                    nextTokenStartPosition - predictedNextTokenStartPosition),
+                                spellChecker,
+                                owner);
+                        if (allTokens.Count > 0)
+                        {
+                            var substToken = GetSpellCheckCleanSubstitutionToken(spellChecker, text, allTokens[allTokens.Count - 1], missingFragment);
+                            if (substToken != null)
+                            {
+                                allTokens[allTokens.Count - 1] = new SpellerSegment(text, substToken.Value, spellChecker, owner);
+                            }
+                        }
+                    }
+                }
+
+                allTokens.Add(
+                    new SpellerSegment(
+                        text,
+                        new WinRTSpellerInterop.TextRange(
+                            nextTokenStartPosition,
+                            (int)tokens[i].SourceTextSegment.Length),
```

Comment:
I think we should introduce the `tokens[i].SourceTextSegment` variable.  This makes the code neat